src/Workspaces/Remote/ServiceHub/Services/CodeAnalysis/CodeAnalysisService_SemanticClassificationCache.cs

```diff
@@ -0,0 +1,295 @@
+            }
+        }
+
+        public Task<SerializableClassifiedSpans> GetCachedSemanticClassificationsAsync(
+            SerializableDocumentKey documentKey, TextSpan textSpan, Checksum checksum, CancellationToken cancellationToken)
+        {
+            return RunServiceAsync(async () =>
+            {
+                using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(out var classifiedSpans);
+
+                await GetOrReadCachedSemanticClassificationsAsync(
+                    documentKey.Rehydrate(), textSpan, checksum, classifiedSpans, cancellationToken).ConfigureAwait(false);
+
+                return SerializableClassifiedSpans.Dehydrate(classifiedSpans);
+            }, cancellationToken);
+        }
+
+        private async Task GetOrReadCachedSemanticClassificationsAsync(
+            DocumentKey documentKey,
+            TextSpan textSpan,
+            Checksum checksum,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            CancellationToken cancellationToken)
+        {
+            // See if we've loaded this into memory first.
+            if (TryGetFromInMemoryCache(documentKey, checksum, classifiedSpans))
+                return;
+
+            // Otherwise, attempt to read in classifications from persistence store.
+            await ReadCachedSemanticClassificationsAsync(
+                documentKey, textSpan, checksum, classifiedSpans, cancellationToken).ConfigureAwait(false);
+
+            UpdateInMemoryCache(documentKey, checksum, classifiedSpans.ToImmutable());
+        }
+
+        private bool TryGetFromInMemoryCache(DocumentKey documentKey, Checksum checksum, ArrayBuilder<ClassifiedSpan> classifiedSpans)
+        {
+            lock (_cachedData)
+            {
+                var data = _cachedData.FirstOrNull(d => d.id == documentKey.Id && d.checksum == checksum);
+                if (data != null)
+                {
+                    classifiedSpans.AddRange(data.Value.classifiedSpans);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void UpdateInMemoryCache(
+            DocumentKey documentKey,
+            Checksum checksum,
+            ImmutableArray<ClassifiedSpan> classifiedSpans)
+        {
+            lock (_cachedData)
+            {
+                // First, remove any existing info for this doc.
+                for (var currentNode = _cachedData.First; currentNode != null; currentNode = currentNode.Next)
+                {
+                    if (currentNode.Value.id == documentKey.Id)
+                    {
+                        _cachedData.Remove(currentNode);
+                        break;
+                    }
+                }
+
+                // Then place the cached information for this doc at the end.
+                _cachedData.AddLast((documentKey.Id, checksum, classifiedSpans));
+
+                // And ensure we don't cache too many docs.
```

Comment:
> So once we do get everything loaded, is the intent that this cache then falls away,

The caches falls away.  This happens in two ways:

1. teh calls to CacheSemanticClassifications will clear it explicitly (since we are fully loaded at that point)
2. the classifier doesn't call into the cache once we are fully loaded.